pkg/sql/schema_changer.go

```diff
@@ -987,10 +1047,102 @@ func (sc *SchemaChanger) RunStateMachineBeforeBackfill(ctx context.Context) erro
 	return nil
 }
 
+// RunStateMachineAfterIndexBackfill moves the state machine forward and
+// wait to ensure that all nodes are seeing the latest version of the
+// table.
+//
+// Adding Mutations in BACKFILLING state move through DELETE ->
+// MERGING.
+func (sc *SchemaChanger) RunStateMachineAfterIndexBackfill(ctx context.Context) error {
+	// Step through the state machine twice:
+	//  - BACKFILLING -> DELETE
+	//  - DELETE -> MERGING
+	log.Info(ctx, "stepping through state machine after index backfill")
+	if err := sc.stepStateMachineAfterIndexBackfill(ctx); err != nil {
+		return err
+	}
+	if err := sc.stepStateMachineAfterIndexBackfill(ctx); err != nil {
+		return err
+	}
+	log.Info(ctx, "finished stepping through state machine")
+	return nil
+}
+
+func (sc *SchemaChanger) stepStateMachineAfterIndexBackfill(ctx context.Context) error {
+	log.Info(ctx, "stepping through state machine")
+
+	var runStatus jobs.RunningStatus
+	if err := sc.txn(ctx, func(
+		ctx context.Context, txn *kv.Txn, descsCol *descs.Collection,
+	) error {
+		tbl, err := descsCol.GetMutableTableVersionByID(ctx, sc.descID, txn)
+		if err != nil {
+			return err
+		}
+		runStatus = ""
+		for _, m := range tbl.AllMutations() {
+			if m.MutationID() != sc.mutationID {
+				// Mutations are applied in a FIFO order. Only apply the first set of
+				// mutations if they have the mutation ID we're looking for.
+				break
+			}
+			idx := m.AsIndex()
+			if idx == nil {
+				// Don't touch anything but indexes
+				continue
+			}
+
+			if m.Adding() {
+				if m.Backfilling() {
+					tbl.Mutations[m.MutationOrdinal()].State = descpb.DescriptorMutation_DELETE_ONLY
+					runStatus = RunningStatusDeleteOnly
+				} else if m.DeleteOnly() {
+					tbl.Mutations[m.MutationOrdinal()].State = descpb.DescriptorMutation_MERGING
+					runStatus = RunningStatusMerging
+				}
+			}
+		}
+		if runStatus == "" || tbl.Dropped() {
+			return nil
+		}
```

Comment:
this and the stepping twice all feels a little brittle. is there a test which resumes between the two steps?

Reply:
I've added a test that randomly pauses before calls to `(*schemaChanger).txn` and then resume the job and makes sure it succeeds. It's a little slow so it only does one pause by default, but locally I've run it after every txn:

```
--- PASS: TestPauseBeforeRandomDescTxn (23.87s)                                                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_1 (1.77s)
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_2 (1.78s)                                        
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_3 (1.74s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_4 (1.80s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_5 (1.91s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_6 (1.72s)                                        
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_7 (1.77s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_8 (1.72s)                                                
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_9 (1.67s)                                                                                                                                                                              
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_10 (1.82s)                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_11 (1.87s)                                               
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_12 (1.94s)
    --- PASS: TestPauseBeforeRandomDescTxn/create_index_pause_at_txn_13 (1.74s)   
```

I think we need to add more assertions that we don't have any unexpected cruft and it would also be nice to add some concurrent operations, but hopefully this gets at the basic problem you are pointing out here.